src/Compilers/CSharp/Portable/Binder/InContainerBinder.cs

```diff
@@ -96,6 +96,79 @@ internal override Imports GetImports(ConsList<Symbol> basesBeingResolved)
+                if ((object)forwardedType != null)
+                {
+                    if (forwardedType.Kind == SymbolKind.ErrorType)
+                    {
+                        DiagnosticInfo diagInfo = ((ErrorTypeSymbol)forwardedType).ErrorInfo;
+
+                        if (diagInfo.Code == (int)ErrorCode.ERR_CycleInTypeForwarder)
+                        {
+                            Debug.Assert((object)forwardedType.ContainingAssembly != null, "How did we find a cycle if there was no forwarding?");
+                            diagnostics.Add(ErrorCode.ERR_CycleInTypeForwarder, location, fullName, forwardedType.ContainingAssembly.Name);
+                        }
+                        else if (diagInfo.Code == (int)ErrorCode.ERR_TypeForwardedToMultipleAssemblies)
+                        {
+                            diagnostics.Add(diagInfo, location);
+                            return null; // Cannot determine a suitable forwarding assembly
+                        }
+                    }
+                    return forwardedType.ContainingAssembly;
+                }
+            }
+            return null;
+        }
+
+        /// <summary>
+        /// Look for a type forwarder for the given type in the containing assembly and any referenced assemblies.
+        /// </summary>
+        /// <param name="name">The name of the (potentially) forwarded type.</param>
+        /// <param name="arity">The arity of the forwarded type.</param>
+        /// <param name="qualifierOpt">The namespace of the potentially forwarded type. If non is provided, will
+        /// try Usings of the current import for egligable namespaces and return the namespace of the found forwarder, 
+        /// if any.</param>
+        /// <param name="diagnostics">Will be used to report non-fatal errors during look up.</param>
+        /// <param name="location">Location to report errors on.</param>
+        /// <returns></returns>
+        /// <remarks>
+        /// Since this method is intended to be used for error reporting, it stops as soon as it finds
+        /// any type forwarder (or an error to report). It does not check other assemblies for consistency or better results.
+        /// </remarks>
+        protected override AssemblySymbol GetForwardedToAssembly(string name, int arity, ref NamespaceOrTypeSymbol qualifierOpt, DiagnosticBag diagnostics, Location location)
+        {
+            name = MetadataHelpers.ComposeAritySuffixedMetadataName(name, arity);
+            var imports = GetImports(basesBeingResolved: null);
+
+            var fullName = qualifierOpt != null && !ReferenceEquals(qualifierOpt, Compilation.GlobalNamespace) ? qualifierOpt.ToDisplayString(SymbolDisplayFormat.QualifiedNameOnlyFormat) + "." + name : name;
+            var result = GetForwardedToAssembly(fullName, diagnostics, location);
+            if (result != null || diagnostics.HasAnyErrors())
+                return result;
```

Comment:
Are we executing this once for each `InContainerBinder` in the `Binder` chain? If so, could this be moved to the caller instead?

Reply:
Yes - I did not understand that there could be multiple InContainerBinders. I changed it now to move the using namespace to a separate method, so only that is executed for each InContainerBinder.